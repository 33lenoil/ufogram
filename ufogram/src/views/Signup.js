import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import register from '../api/registration';
import getUsers from '../api/user';

function Signup() {
    // if account does not exist, retry or jump to registration page
    // if password incorrect, retry, block if wrong for 3 times
    // if password correct, jump to Main_view

    const navigate = useNavigate();

    let [username, setUsername] = useState('');
    let [password, setPassword] = useState('');

    const handleSignup = async () => {
        
        console.log('User info', username, password);
        const response = await getUsers();
        for (let i = 0; i < response.data.length; i++) {
                    
            if (response.data[i].username === username) {
                return;
            }
        }
        try {
            const status = await register(username, password);
            console.log('Status', status);
            if (status === 201) {
                navigate('/login');
            }
        } catch (err) {

        }
        
    }

    const handleLogin = () => {
        navigate('/login');
    }

    function handleUsernameChange(unameEvent) {
        setUsername(unameEvent.target.value);
    }

    function handlePasswordChange(passwdEvent) {
        setPassword(passwdEvent.target.value);
    }

    return (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
            <div style={{display: "flex", justifyContent: "center", position: "fixed", width: "100%", background: "#8769b6"}}>
                <h1>UFOgram Signup</h1>
            </div>
            <div style={{display: "flex", width: "12%", marginTop: "100px", flexDirection: "column"}}>
                <label htmlFor='Username'>Username: </label>
                <input data-testID='uname' type="text" name="Username" value={username} onChange={handleUsernameChange}/>
                <label htmlFor='Password'>Password: </label>
                <input type="password" name="Password" value={password} onChange={handlePasswordChange}/>
                <button id='sign_up' title="Sign up" onClick={handleSignup}>Signup</button>
                <button title="Log in" onClick={handleLogin} style={{color: "#808080"}}>Login</button>
            </div>
        </div>
    );
}

export default Signup;

