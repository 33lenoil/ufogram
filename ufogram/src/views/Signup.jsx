import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import register from '../api/registration';
import getUsers from '../api/user';

function Signup() {
  const navigate = useNavigate();

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const handleSignup = async () => {
    // console.log('User info', username, password);
    const response = await getUsers();
    for (let i = 0; i < response.data.length; i += 1) {
      if (response.data[i].username === username) {
        return;
      }
    }
    try {
      const status = await register(username, password);
      // console.log('Status', status);
      if (status === 201) {
        navigate('/login');
      }
    } catch (err) {
      // console.log('Register error');
    }
  };

  const handleLogin = () => {
    navigate('/login');
  };

  function handleUsernameChange(unameEvent) {
    setUsername(unameEvent.target.value);
  }

  function handlePasswordChange(passwdEvent) {
    setPassword(passwdEvent.target.value);
  }

  return (
    <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
      <div style={{
        display: 'flex', justifyContent: 'center', position: 'fixed', width: '100%', background: '#8769b6',
      }}
      >
        <h1>UFOgram Signup</h1>
      </div>
      <div style={{
        display: 'flex', width: '12%', marginTop: '100px', flexDirection: 'column',
      }}
      >
        <label htmlFor="Username">
          Username:
          <input type="text" name="Username" value={username} onChange={handleUsernameChange} />
        </label>
        <label htmlFor="Password">
          Password:
          <input type="password" name="Password" data-testid="passwordBox" value={password} onChange={handlePasswordChange} />
        </label>
        <button type="button" id="sign_up" title="Sign up" onClick={handleSignup}>Signup</button>
        <button type="button" title="Log in" onClick={handleLogin} style={{ color: '#808080' }}>Login</button>
      </div>
    </div>
  );
}

export default Signup;
