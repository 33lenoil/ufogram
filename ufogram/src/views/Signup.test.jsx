/**
 * @jest-environment jsdom
 */

import React, { useState } from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Signup from './Signup';
import { act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import * as router from 'react-router-dom';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

beforeEach(() => {
    expect(router)
    });

    const mockAxios = new MockAdapter(axios);

  const users = [
    {
      username: 'zairuiy',
      password: '1234567',
      following: [
        {
          username: 'lionelhu',
          id: 2,
        },
      ],
      followers: [
        {
          username: 'lionelhu',
          id: 2,
        },
      ],
      id: 1,
    },
    {
      username: 'lionelhu',
      password: '1234567',
      following: [
        {
          username: 'zairuiy',
          id: 1,
        },
      ],
      followers: [
        {
          username: 'zairuiy',
          id: 1,
        },
      ],
      id: 2,
    },
  ];

  mockAxios.onGet('http://localhost:3000/Users').reply(200, users);
  mockAxios.onPost('http://localhost:3000/Users').reply(201, users);

    const mockedNavigate = jest.fn();

    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        // useNavigate: () => ({ navigate: mockedNavigate})
        useNavigate: () => mockedNavigate,
    }));

test('renders title', () => {
    render(
      <Router>
          {/* <Routes>
              <Route path="/signup" element={<Signup />} />
          </Routes> */}
          <Signup />
      </Router>
    );
    const linkElement = screen.getByText(/UFOgram/);
    expect(linkElement).toBeInTheDocument();
  });

test('renders username label', () => {
  render(
    <Router>
        {/* <Routes>
            <Route path="/signup" element={<Signup />} />
        </Routes> */}
        <Signup />
    </Router>
  );
  const linkElement = screen.getByText(/Username:/);
  expect(linkElement).toBeInTheDocument();
});

test('renders password label', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Signup />} />
            </Routes> */}
            <Signup />
        </Router>
      );
      const linkElement = screen.getByText(/Password:/);
    expect(linkElement).toBeInTheDocument();
});

test('renders signup button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Signup />} />
            </Routes> */}
            <Signup />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Signup/
  })
  expect(linkElement).toBeInTheDocument();
});

test('renders login button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Signup />} />
            </Routes> */}
            <Signup />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Login/
  })
  expect(linkElement).toBeInTheDocument();
});

/**
 * Snapshot Testing
 */

test('the component matches the snapshot', () => {
  const component = renderer.create(<Router>
    {/* <Routes>
        <Route path="/signup" element={<Signup />} />
    </Routes> */}
    <Signup />
</Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test the login button', async () => {
    act(() => {
        render(
            <MemoryRouter initialEntries={['/signup']}> {/* Set the initial route */}
              <Signup />
            </MemoryRouter>
          );
    })
    // render(
    //   <MemoryRouter initialEntries={['/login']}> {/* Set the initial route */}
    //     <Login />
    //   </MemoryRouter>
    // );

    const loginButton = screen.getByText('Login');
    act(() => {fireEvent.click(loginButton)});

    // check if you have successfully navigated to the expected page
    // const newPageElement = screen.getByText('Login');

    // expect(newPageElement).toBeInTheDocument();
    // expect(navigate).toHaveBeenCalledWith('/signup');
    await waitFor(() => {
        expect(mockedNavigate).toHaveBeenCalledWith('/login');
    });
});

test('Tests signup button', async () => {
    act(() => {
        render(
            <MemoryRouter initialEntries={['/signup']}> {/* Set the initial route */}
              <Signup />
            </MemoryRouter>
          );
    })
    // render(
    //   <MemoryRouter initialEntries={['/login']}> {/* Set the initial route */}
    //     <Login />
    //   </MemoryRouter>
    // );

    const usernameBox = screen.getByRole('textbox')
    // const usernameBox = screen.getByDisplayValue('usernameBox');
  
    act(() => {
        userEvent.type(usernameBox, 'test11');
    })

    // const passwordBox = screen.getByRole('textbox')
    const passwordBox = screen.getByTestId('passwordBox');
  
    act(() => {
        userEvent.type(passwordBox, '1234567');
    })

    const signupButton = screen.getByText('Signup');
    act(() => {fireEvent.click(signupButton)});

    // check if you have successfully navigated to the expected page
    // const newPageElement = screen.getByText('Signup');

    // expect(newPageElement).toBeInTheDocument();
    await waitFor(() => {
        expect(mockedNavigate).toHaveBeenCalledWith('/login');
    });
    // expect(mockedNavigate).toHaveBeenCalledWith('/main');
    // await waitFor(() => expect(window.location.href).toContain('/main'));
});

/**
 *  Testing
 */

// test('welcome is displayed after click the registration button', async () => {
//   render(<Signup />);
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByLabelText(/welcome/i);
//   expect(welcomeElement).toBeInTheDocument();
// });

// test('Username is displayed after clicking on registration button', async () => {
//   render(<Signup />);
//   const usernameElement = screen.getByTestId('uname');
//   await userEvent.type(usernameElement, 'lili');
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByText(/lili/i);
//   expect(welcomeElement).toBeInTheDocument();
// });
