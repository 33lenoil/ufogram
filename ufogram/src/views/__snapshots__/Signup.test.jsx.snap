// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-evenly",
    }
  }
>
  <div
    style={
      Object {
        "background": "#8769b6",
        "display": "flex",
        "justifyContent": "center",
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <h1>
      UFOgram Signup
    </h1>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginTop": "100px",
        "width": "12%",
      }
    }
  >
    <label
      htmlFor="Username"
    >
      Username:
      <input
        name="Username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="Password"
    >
      Password:
      <input
        data-testid="passwordBox"
        name="Password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <button
      id="sign_up"
      onClick={[Function]}
      title="Sign up"
      type="button"
    >
      Signup
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "#808080",
        }
      }
      title="Log in"
      type="button"
    >
      Login
    </button>
  </div>
</div>
`;
